### PR TITLE
docs: accounts

### PR DESCRIPTION
**Pull request by enitrat**
Time spent on this PR: 0.2d

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Resolves #<Issue number>

## What is the new behavior?

-
-
-

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1056)

### CODE REVIEW

File: docs/general/accounts.md [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+Starknet infrastructure. If you own the keys to an Ethereum address, you can use
+the `deploy_externally_owned_account` entrypoint to deploy a corresponding
+Starknet account contract, whose address is deterministic based on the Ethereum
+address.
+
+Although our backend makes no distinction between EOA and CA, the validation
+logic will reject any transaction that originates from a contract that has EVM
+code. In the (extremely unlikely) event that a user sends a transaction from a
+contract account after mining the private key corresponding to an EOA, the
+transaction will be rejected.
```
Comment (ClementWalter) [inline docs/general/accounts.md:22]: same behavior as in the EVM right? I mean, if you have the private key of a CA, tx will be rejected as well. Just make sure that it's not possible to understand that Kakarot as a flow/issue not existing in an original EVM

File: docs/general/accounts.md [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+## Account storage
+
+Account contracts store the following information:
+
+- `storage`: A mapping from 32-byte keys to 32-byte values. This is used to
+  store the values stored in the EVM account storage.
+- `nonce`: A 32-byte value representing the number of transactions sent from the
```
Comment (ClementWalter) [inline docs/general/accounts.md:30]: it's a felt right?

File: docs/general/accounts.md [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+  account. Even though EOAs do not use the stored nonce for validation, it is
+  still stored in the account contract to ensure the stored nonce always matches
+  the protocol nonce, which is incremented by one by the sequencer for each
+  transaction.
+- `bytecode`: The EVM bytecode of the account. This is only used for CA
+  accounts, as EOAs do not have bytecode.
```
Comment (ClementWalter) [inline docs/general/accounts.md:36]: Contract Accounts accounts is a bit redundant

File: docs/general/accounts.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+- `version`: A 9-digit integer representing the version of the account contract
```
Comment (ClementWalter) [inline docs/general/accounts.md:37]: add the `xxx.yyy.zzz` comment as in the contract

File: docs/general/accounts.md [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+The account contract class also provides entrypoints related to account
+management: the current implementation hash, the version of the account contract
+class, and the Ethereum address associated with the account. These entrypoints
+are used to verify the account state and to upgrade the account contract class.
+
+The account contract class also has `__validate__` and `__execute__`
+entrypoints, which are used to process transactions originating from EOAs. These
+entrypoints are called by the sequencer when a transaction is submitted to the
+sequencer. The `__validate__` entrypoint first checks if the transaction is
+valid, and the `__execute__` entrypoint processes the transaction if it is
```
Comment (ClementWalter) [inline docs/general/accounts.md:62]: maybe a link to starknet doc about accounts?

File: docs/general/accounts.md [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+   owner of the account to upgrade the account contract class to a new version.
+2. Before upgrading an account's class, the system checks whether the new class
+   supports the account interface using
+   [SNIP-5](https://github.com/starknet-io/SNIPs/blob/main/SNIPS/snip-5.md),
+   Cairo's specific implementation of
+   [EIP-165](https://eips.ethereum.org/EIPS/eip-165), to avoid breaking the
+   account's functionality.
+3. Accounts have a `version` field that stores the version of the account
+   contract class. This field is used to check if the account contract class can
+   be upgraded. It is not possible to downgrade the account contract class.
```
Comment (ClementWalter) [inline docs/general/accounts.md:104]: why not possible to downgrade?

File: docs/general/accounts.md [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+4. The `version` field should be a 9-bit integer, where the first 3 bits
```
Comment (ClementWalter) [inline docs/general/accounts.md:105]: 9 digits

**Review by ClementWalter — CHANGES_REQUESTED** [review]
some minor comments

File: docs/general/accounts.md (hunk 2 above)
Comment (enitrat) [reply docs/general/accounts.md:30]: Actually it's a u64 value, represented in a felt sorry

File: docs/general/accounts.md [hunk 8]
```diff
@@ -0,0 +1,112 @@
+# Accounts
+
+While the EVM defines two types of accounts (EOAs and CAs), Starkent only has
```
Comment (ClementWalter) [inline docs/general/accounts.md:3]: typo Starkent

File: docs/general/accounts.md [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+user sends a transaction from a contract account after mining the private key
+corresponding to an EOA, the transaction will be rejected.
+
+## Account storage
+
+Account contracts store the following information:
+
+- `storage`: A mapping from 32-byte keys to 32-byte values. This is used to
+  store the values stored in the EVM account storage.
+- `nonce`: A 64-bits value representing the number of transactions sent from the
```
Comment (ClementWalter) [inline docs/general/accounts.md:30]: why only 64 bits?

File: docs/general/accounts.md [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+  account. Even though EOAs do not use the stored nonce for validation, it is
+  still stored in the account contract to ensure the stored nonce always matches
+  the protocol nonce, which is incremented by one by the sequencer for each
+  transaction.
+- `bytecode`: The EVM bytecode of the account. This is only used for CAs, as
+  EOAs do not have bytecode.
+- `version`: A 9-digit integer representing the version of the account contract,
+  in the format `major.minor.patch` (3 digits each). class.
```
Comment (ClementWalter) [inline docs/general/accounts.md:38]: extra word

File: docs/general/accounts.md (hunk 9 above)
Comment (enitrat) [reply docs/general/accounts.md:30]: Starknet protocol limitation. In reality the nonce is also 64 bits on EVM clients, and attempting to send a tx with nonce >= 2**64 fails the EVM protocol rules.

File: docs/general/accounts.md (hunk 6 above)
Comment (ClementWalter) [reply docs/general/accounts.md:106]: update this now

File: docs/general/accounts.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+
+## Deterministic mapping of addresses
+
+The account model allows us to compute the address of an account contract from
+an Ethereum address. The address of a starknet contract is computed from the
+following inputs:
+
+- The hash of the starknet contract class.
+- A salt, for which we use the Ethereum address.
+- Eventual constructor arguments. We use the address of the Kakarot contract and
```
Comment (ClementWalter) [inline docs/general/accounts.md:77]: + implementation class hash

File: docs/general/accounts.md [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+
+This system ensures a deterministic calculation of the starknet contract address
+from an Ethereum address. However, this requires all inputs to be immutable, as
+changing any input will result in a different starknet contract address. This
+requires a deployment process, where we first deploy a contract using a
+restricted class that will always stay immutable (`uninitialized_account`), and
+then upgrade the class of the contract to the desired account class using the
+most up-to-date class. Anyone can deploy an account contract using the
+`deploy_externally_owned_account` entrypoint from Kakarot, or do the same
+manually using the `deploy` syscall.
```
Comment (ClementWalter) [inline docs/general/accounts.md:89]: update once #1065 is merged

File: docs/general/accounts.md [hunk 13]
```diff
@@ -0,0 +1,175 @@
...
+
+The deployment process is illustrated as follows:
+
+```mermaid
+sequenceDiagram
+    Anyone ->>+ UninitializedAccount: deploy_syscall
+    UninitializedAccount->>+Kakarot: get_account_contract_class_hash()
+    Kakarot-->>-UninitializedAccount: account_contract_class_hash
+    UninitializedAccount->>+AccountContractClass: library_call_initialize(kakarot_address,evm_address,account_contract_class_hash)
+    AccountContractClass ->> AccountContractClass: set_owner(kakarot_address)
```
Comment (ClementWalter) [inline docs/general/accounts.md:107]: non consistent spacing around arrows

File: docs/general/accounts.md (hunk 11 above)
Comment (enitrat) [reply docs/general/accounts.md:77]: as part of the constructor argument?

**Review by ClementWalter — APPROVED** [review]

**Pull request merged**